test/rematch.jl

```diff
@@ -18,32 +18,248 @@ assertion_error = (VERSION >= v"0.7.0-DEV") ? LoadError : AssertionError
         y
     end
 
-    # test basic match
-    let x = nothing
-        @test (@match Foo(x,2) = Foo(1,2)) == Foo(1,2)
-        @test x == 1
+    @eval begin
+        # test basic match
+        let x = nothing
+            @test (@match Foo(x,2) = Foo(1,2)) == Foo(1,2)
+            @test x == 1
+        end
+
+        # variables not bound if match fails
+        let x = nothing
+            @test_throws MatchFailure @match Foo(x, 3) = Foo(1,2)
+            @test x == nothing
+        end
+
+        # doesn't overwrite variables in outer scope
+        let x = nothing
+            @test (@match Foo(1,2) begin
+                Foo(x,2) => x
+            end) == 1
+            @test x == nothing
+        end
+
+        # variables not bound if guard fails
+        let x = nothing
+            @test_throws MatchFailure @match Foo(1,2) begin
+              Foo(x, 2) where x != 1 => :ok
+            end
+            @test x == nothing
+        end
     end
+end
 
-    # variables not bound if match fails
+@testset "Match using extractors" begin
+    function sub1(x) x+1 end
+
+    # sub1(4) == 3
     let x = nothing
-        @test_throws MatchFailure @match Foo(x, 3) = Foo(1,2)
+        @test (@match 3 begin
+           ~sub1(x) => x
+        end) == 4
```

Comment:
I don't understand why this returns `4`. I could understand if it returned `2` or `3`, but i don't quite understand how it returns `4`.

The definition of `sub1(x)` is `sub1(x) = x+1`. So I take `@match 3 begin   sub1(x) => x  end` to mean "if `3` matches `sub1(x)` for some `x`, return that `x`." And in this case, `3` would match `sub1(2)`, so i'd expect it to return `2`?

Also, i'm confused why `x+1` is called `sub1` - i'd have called it `add1`?

Reply:
Ohhhhhhhh after reading some examples, i now understand that in the above match pattern, the `x` doesn't refer to the input, it refers to the _output_.

So `~sub1(x)` means `x == sub1(3)`, meaning `x == 4`.

This was clearer to me after reading the Cons example in the README.


---------------

Perhaps we can find a better syntax that doesn't look like a function application? This would also help with picking a disambiguating syntax from struct constructors.

What do you think about something like this?:
```julia
@test (@match 3 begin
    sub1(~)(x) => x
end) == 4

@match [1,2,3] begin
    Cons(~)(x, xs) => @assert x == 1 && xs == [2,3]
end
```
Where the `~` represents the object being matched on.

Or instead of the `~`, maybe `Cons(_)(x, xs)`? Or `Cons()(x, xs)`? Or maybe something like `Cons()->(x, xs)`, or some other similar but illegal syntax?

Reply:
> I don't understand why this returns `4`. I could understand if it returned `2` or `3`, but i don't quite understand how it returns `4`.

The idea is that `3 == sub1(4)`. Don't think of `sub1` as a normal function; it's basically the inverse of a constructor.
Maybe we can change the name to `unapply_sub1` or something like that (in Scala extractor methods are all named `unapply`, the inverse of `apply`). This would avoid overloading issues if there's a unary constructor also.

> Perhaps we can find a better syntax that doesn't look like a function application? This would also help with picking a disambiguating syntax from struct constructors.
> 
> What do you think about something like this?:
> 
> ```julia
> @test (@match 3 begin
>     sub1(~)(x) => x
> end) == 4
> 
> @match [1,2,3] begin
>     Cons(~)(x, xs) => @assert x == 1 && xs == [2,3]
> end
> ```

I don't see this as any better than `~Cons(x,xs)`.

Reply:
Ah, but renaming the function to `unapply_X` breaks parameterized extractors like `Re` in the tests.

Reply:
Nevermind, easy to fix. Extractors are named `unapply_Foo` and the pattern is `~Foo`. Pushed.

Reply:
After using it a bit, I'm not really sure I like the `unapply_` prefix (or any prefix). Best to just have the writer of the extractor function call it whatever they want, avoiding confusion about which function gets called.

Reply:
> I don't see this as any better than `~Cons(x,xs)`.

The reason I prefer `Cons(~)(x,xs)` is because it prevents parsing the above as a _two argument function call to `Cons`_, which is what it really looks like.

Instead, I want to somehow indicate that the syntax is matching on a two-argument destructured _return value_ from calling `Cons` _on the match variable_. I was imagining the `Cons(~)` syntax to imply "calling `Cons` with the match variable", and then `Cons(~)(x,xs)` implying "the return value from that function can be captured by destructuring into two variables."

Another option might be something like `(x,xs) = Cons(~)`?

Basically, I just want to prevent the initial parse looking like the destructed output variables are actually the inputs to the function, since when they're both arity-1, as in the `sub1` function, it was difficult to understand which variable is the input and which is the output.

Does that make sense? :)

This seems like a really cool feature! I'm sorry to hold it up bikeshedding. Please take this as a positive: this is really cool and I want to help make it intuitive and easy to use! :)

Reply:
Sorry we left this PR lagging for _forever_.

Please feel free to just merge it now, and we can always change this in later iterations if we want to!

After rereading the above discussion, i would like to propose that latest suggestion one more time; i actually think it's fairly nice:

```julia
@test (@match 3 begin
    (x = sub1(~)) => x
end) == 4

@match [1,2,3] begin
    ((x, xs) = Cons(~)) => @assert x == 1 && xs == [2,3]
end
```

This syntax makes the behavior more obvious, IMO:
- call the supplied function on the match variable, and then
- do a capture match on the return value.

----------------------

BUT it seems to me like you are implementing a well known feature in other pattern-matching systems, and I am _not at all_ familiar with those, so please feel free to completely ignore my input. :)

I'd like to get out of the game of causing drama on this repo! ❤️ 😅